pkg/manifests/rte/rte.go

```diff
@@ -45,16 +45,19 @@ type Manifests struct {
 	// internal fields
 	plat           platform.Platform
 	serviceAccount string
+	// DaemonSet which not processed by Manifests.Update()
+	initialDaemonSet *appsv1.DaemonSet
 }
 
 func (mf Manifests) Clone() Manifests {
 	ret := Manifests{
 		plat:           mf.plat,
 		serviceAccount: mf.serviceAccount,
 		// objects
-		Role:        mf.Role.DeepCopy(),
-		RoleBinding: mf.RoleBinding.DeepCopy(),
-		DaemonSet:   mf.DaemonSet.DeepCopy(),
+		Role:             mf.Role.DeepCopy(),
+		RoleBinding:      mf.RoleBinding.DeepCopy(),
+		DaemonSet:        mf.DaemonSet.DeepCopy(),
+		initialDaemonSet: mf.initialDaemonSet.DeepCopy(),
```

Comment:
this is workarounding a deeper bug. `mf.DaemonSet` is indeed supposed to be the initial pristine copy. We should never mutate that copy. This is why `Update` begins with `ret := mf.Clone()`.
Where in the flow we mutate the original reference of the DaemonSet?

Reply:
hmm interesting let me look deeper.

Reply:
Damn! :) https://github.com/openshift-kni/numaresources-operator/blob/main/controllers/numaresourcesoperator_controller.go#L117

Reply:
nevertheless, the bug is real. The shell expansion in the DS command happens only once, and this is not the original intent.

Reply:
I couldn't come up with a better patch that addresses the shell expansion specifically.
If you have an idea I'm open to suggestions.

Reply:
After more closer inspection, `Clone()` and `Update()` work as intended. `DeepCopy()` is doing the right thing.
The bug is real, but is in how the manifests are consumed and not in how we generate and update them.

The only real bug here is in the naming: most of the functions the original author ( :) ) called `Update` aren't updating, rather transforming the manifests. "update" is typically in-place, and here we go to some great extent to avoid in-place transformations.

In the future we should rename the code with more explicative names.

Reply:
I agree. This bug is more about wrong using than wrong implementation.

Anyway let me know what you prefer.
Just merging the unittest from this PR and remove the rest, is also an option.

Reply:
The test addition is a good idea, but let's file a new PR later in the future just adding tests.

Reply:
Let's add tests to verify that the behaviour of `Clone()` and `Update()` - which should never alter the original reference, but always return a mutated copy - is indeed correct. Of course let's do in a new PR.

Reply:
Sure. I Will issue PR soon.